src/typing/codegen.ml

```diff
@@ -0,0 +1,298 @@
+(**
+ * Copyright (c) 2013-present, Facebook, Inc.
+ * All rights reserved.
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the "flow" directory of this source tree. An additional grant
+ * of patent rights can be found in the PATENTS file in the same directory.
+ *
+ *)
+
+let spf = Printf.sprintf
+
+type codegen_env = {
+  buf: Buffer.t;
+  declared_classes: string IMap.t;
+  mutable next_declared_class_name: int;
+  flow_cx: Context.t;
+  indent: string;
+  tparams: Type.typeparam list;
+  applied_tparams: Type.t list;
+}
+
+let add_str str env = Buffer.add_string env.buf str; env
+let add_tparams tparams env =
+  {env with tparams;}
+let add_applied_tparams applied_tparams env =
+  {env with applied_tparams;}
+let find_props tmap_id env = Flow_js.find_props env.flow_cx tmap_id
+let next_declared_class_name env =
+  let id = env.next_declared_class_name in
+  env.next_declared_class_name <- id + 1;
+  id
+let resolve_type t env = Flow_js.resolve_type env.flow_cx t
```

Comment:
On mobile so forgive an unresearched comment: I think Flow_js.resolve_type will return `any` for open tvars, so ideally this should happen after merge. Does it?